TFUtil.py

```diff
@@ -8517,3 +8517,177 @@ def get_non_deterministic_ops_from_graph():
     # elif ... more non det ops to be added
 
   return non_det_ops
+
+
+def sum_rows(x):
+  """
+  Returns a vector summing up each row of the matrix x.
+
+  This is mostly a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1427
+
+  :param tf.Tensor x: input tensor
+  :return: Vector, which rows are summed along axis 1 with shape [shape(x)[0]]
+  :rtype: tf.Tensor
+  """
+  # sum_rows(x) is equivalent to math_ops.reduce_sum(x, 1) when x is
+  # a matrix.  The gradient of _sum_rows(x) is more efficient than
+  # reduce_sum(x, 1)'s gradient in today's implementation. Therefore,
+  # we use _sum_rows(x) in the nce_loss() computation since the loss
+  # is mostly used for training.
+  cols = tf.shape(x)[1]
+  ones_shape = tf.stack([cols, 1])
+  ones = tf.ones(ones_shape, x.dtype)
+  return tf.reshape(tf.matmul(x, ones), [-1])
+
+
+def compute_sampled_logits(weights,
+                           biases,
+                           labels,
+                           inputs,
+                           num_sampled,
+                           num_classes,
+                           num_true=1,
+                           sampled_values=None,
+                           subtract_log_q=True,
+                           remove_accidental_hits=False,
+                           partition_strategy="mod",
+                           name=None,
+                           seed=None):
+  """Helper function for nce_loss and sampled_softmax_loss functions.
+  Computes sampled output training logits and labels suitable for implementing
+  e.g. noise-contrastive estimation (see nce_loss) or sampled softmax (see
+  sampled_softmax_loss).
+  Note: In the case where num_true > 1, we assign to each target class
+  the target probability 1 / num_true so that the target probabilities
+  sum to 1 per-example.
+
+  This is mostly a copy of
+    https://github.com/tensorflow/tensorflow/blob/e19c354920c3b246dda6598229210a582caaa1a9/tensorflow/python/ops/nn_impl.py#L1440
+
+  :param weights tf.Tensor|[tf.Tensor]: A `Tensor` of shape `[num_classes, dim]`, or a list of `Tensor`
```

Comment:
The syntax is wrong. It should be `:param tf.Tensor|list[tf.Tensor| weights: ...`.